core/new-gui/src/app/workspace/service/joint-ui/joint-ui.service.ts

```diff
@@ -0,0 +1,253 @@
+ *  - adding input and output ports to the box based on the operator metadata
+ *  - changing the look of the ports
+ *  - adding a new delete button and the callback function of the delete button,
+ *      (original JointJS element doesn't have a built-in delete button)
+ *
+ * @author Henry Chen
+ * @author Zuozhi Wang
+ */
+@Injectable()
+export class JointUIService {
+
+  private operators: OperatorSchema[] = [];
+
+  /**
+   * This variable defines the svg property
+   * of a delete button
+   */
+  public readonly deleteButtonPath =
+  'M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41' +
+  ' 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2' +
+  ' 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z';
+
+  /**
+   * This variable defines the wrapper around
+   * the button, deciding the size and other styles
+   * for the delte button.
+   */
+  public readonly deleteButtonSVG =
+  `<svg class="delete-button" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
+    <path d="M0 0h24v24H0z" fill="none" pointer-events="visible" />
+    <path d="${this.deleteButtonPath}"/>
+  </svg>`;
+
+  /**
+   * This variable defines the handle (the square at the end)
+   * of the source operator for a link
+   */
+  public readonly sourceOperatorHandle = 'M 0 0 L 0 8 L 8 8 L 8 0 z';
+
+  /**
+   * This variable defines the handle (the arrow at the end)
+   * of the target operator for a link
+   */
+  public readonly targetOperatorHandle = 'M 12 0 L 0 6 L 12 12 z';
+
+  constructor(
+    private operatorMetadataService: OperatorMetadataService
+  ) {
+    this.operatorMetadataService.getOperatorMetadata().subscribe(
+      value => this.operators = value.operators
+    );
+
+    this.setupCustomJointjsModel();
+  }
+
+  /**
+   * Gets the JointJS UI Element Object based on OperatorType OperatorID.
+   *
+   * The JointJS Element could be added to the JointJS graph to let JointJS display the operator accordingly.
+   * The function first check if the operatorType exists in the metadata, if it doesn't,
+   * the program will throw an error. Then it creates a JointJS shape created from the
+   * setupCustomJointjsModel() function called in the constructor above.
+   *
+   * @param operatorType the type of the operator
+   * @param operatorID the ID of the operator, the JointJS element ID would be the same as operatorID
+   * @param xPosition the topleft x position of the operator element (relative to JointJS paper, not absolute position)
+   * @param yPosition the topleft y position of the operator element (relative to JointJS paper, not absolute position)
+   *
+   * @returns JointJS Element
+   */
+  public getJointjsOperatorElement(
+    operatorType: string, operatorID: string, xPosition: number, yPosition: number
+  ): joint.dia.Element {
+
+    const operatorSchema = this.operators.find(op => op.operatorType === operatorType);
```

Comment:
how do we know that the call in contructor has completed and this.operators is not empty.

Reply:
ok. so am I correct in saying that we assume that when drag and drop is used, the user will take some time to do that, and till then the operatorMetadataService.getOperatorMetadata() will finish and bring in the metadata? And if it hasn't finished and the user does something, the action will throw an error.

I realize that might not happen because if the service doesn't succeed in bringing the metadata, there will be nothing for the user to drag and drop. Is that correct?

Reply:
However, still we might have a to-do item which we can take up later, to do something like blurring out the screen and showing a loading sign till some initialization is taking place. At that time, we don't allow users to do anything

Reply:
@avinash0161 we don't need that (at least for now). The operator panel won't have show operators initially because the data hasn't arrived yet. And users have nothing to drag.